### PR TITLE
cryptsetup: support keyfile-timeout for using a device as the key file

### PR DESCRIPTION
**Pull request by yan12125**
Closes https://github.com/systemd/systemd/issues/21993

---

Didn't update tests as I cannot find existing tests for cryptsetup codes.

### CODE REVIEW

File: src/cryptsetup/cryptsetup-generator.c [hunk 1]
```diff
@@ -227,7 +227,7 @@ static int generate_device_umount(const char *name,
         return 0;
 }
 
-static int print_dependencies(FILE *f, const char* device_path) {
+static int print_dependencies(FILE *f, const char* device_path, const char* timeout_value, bool canfail) {
         int r;
```
Comment (bluca) [inline src/cryptsetup/cryptsetup-generator.c:232]: `assert(!canfail || timeout_value)`

**Comment by poettering** [comment]
Looks OK. 

But looking at this, I really dislike the fact this so far has been passing around the timeout as string. Instead, this should parse the timeout with parse_usec(), then pass a usec_t around, and make a string from it as FORMAT_TIMESPAN()...

Taking unvalidated data as input and propagating it as-is to the output, sucks. We realy should parse/validate it, and then regenerate it in uniform way.

Anyway, this is an issue predating this PR. Hence I can't blame you for this, and the PR is good. But if you want extra points, it would be excellent to rework this to parse the timeout early on, and then format it back to a string where needed.

**Comment by yan12125** [comment]
Yeah, indeed passing unparsed strings around is not that good. Looking into time-related functions in systemd, I feel that doing it correctly needs changes more than a few lines. I prefer such changes to be in another pull request.

**Review by bluca — APPROVED** [review]
That's fair, let's merge this

**Comment by yan12125** [comment]
Thank you all for reviewing!

**Pull request merged**